### PR TITLE
Bailey new personal max backend

### PR DESCRIPTION
**Pull request by BaileyM96**
# Description
18. (PRIORITY High) Bailey/Jae: Fix the New Max Personal Record award.
The badge should just update the number next to it AND on it, and not be awarded again in the list shown below.
Every time the record is broken though, it should still show up in the New Badges section of the person’s Dashboard

## Related PRS (if any):
None

## Main changes explained:
- Added badgeCollection of only Personal Max badges to badgeOfType since badgeOfType was defined but was assigned no value.
- Created an empty array called duplicateBadges for the duplicateBadges to be pushed into and removed without removing all Personal Max Badges.
- Improved the removeDupBadge functionality to look for the objectId of the badge rather than just the badgeId.

## How to test:
1. check into current branch
2. do `npm install`  to run this PR locally
3. Navigate to cronjobs folder→userProfileJobs→change the cronjob from '1 0 * * * ' to '* * * * *' so it runs every minute.
4. Open MongoDb compass and find your user profile that you want to test and change personalBestMaxHrs to a value that is greater than 10→change lastWeekTangibleHrs to match the personalBestMaxHrs.
5. do `npm run build && npm start` This ensures the changes for the cronjob go into affect so it will run right away.
6. Login as your test account and Empty cache and hard reload so the changes appear on the page under Badges for the front end.

## Screenshots or videos of changes:
Walkthrough of how to test the back end.
[Personal Max Testing Guide.webm](https://github.com/OneCommunityGlobal/HGNRest/assets/108101478/09629d0b-9288-429d-b842-fa6fe4d93fb0)

The fields to change in the MongoDB database
Showing what to change in MongoDB

## Note:
The badge only awards on Sundays, this testing method just helps award the badge right away to make sure the badge does not duplicate. 

When awarding the badge right away it will not appear under new badges until Sunday and the count of the badges inside the report badges will not update until Sunday as well.

### CODE REVIEW

**Review by robertoooc — APPROVED** [review]
Hey @BaileyM96, I tested this pr and found everything to work well. Changing the cronjob's running period allowed the badges to remove duplicates and update the personal max hrs appropriately. It took a few page refreshes to see it work but that makes sense from the cronjob running every minute.

One thing I additionally did was comment out 527 -531 in the userProfileController.js file, this was just so caching wouldn't get in the way of the immediate update. 

Only thing I would suggest is to remove unneeded console.logs.


https://github.com/OneCommunityGlobal/HGNRest/assets/117053202/faf3ff34-6e8a-4d20-adfd-1a57f8dbbddb

**Review by luisarevalo21 — APPROVED** [review]
Hello @BaileyM96. 

I tested your PR and works as intended. Great work!
I modified, the personalBestMaxHrs and lastWeekTangibleHrs on the backend to 20. I then ran `npm run build && npm start`
Lastly, I did a hard refresh and showed up immediately, however, i still ran empty cache and worked as intended. The badge updated to 22 on the dashboard page and in the user's profile as well. 


https://github.com/OneCommunityGlobal/HGNRest/assets/22454953/142b66a7-a23d-4763-9e22-c2fc5b282456

Regardless, I agree with Roberto to remove any console.logs. 
Overall great work!

**Review by ilyaflaks — APPROVED** [review]
I have reviewed this PR and it works as intended
PR619-13hrs
Then this user beat their record. Old badge got removed and a new one (15 hrs) was awarded:
PR619-15hrs
Great work!

File: src/controllers/userProfileController.js
Comment (ptrpengdev) [inline src/controllers/userProfileController.js:1]: Should remove console.log statement.

File: src/controllers/userProfileController.js [hunk 1]
```diff
@@ -51,6 +51,7 @@ async function ValidatePassword(req, res) {
 }
 
 const userProfileController = function (UserProfile) {
+  console.log('Hello')
```
Comment (ptrpengdev) [inline src/controllers/userProfileController.js:54]: Should remove console.log statement.

**Review by jerryren527 — APPROVED** [review]
Hi @BaileyM96 ,

Great job on this task! Your code changes seem to work as intended. After changing personalBestMaxHrs to a value greater than 10, running the backend server, and hard refreshing, the changes to the personal max hours appear on the test account's profile page.


https://github.com/OneCommunityGlobal/HGNRest/assets/55968519/a25476a0-ec01-4aea-909d-1f920d58a469

**Review by ptrpengdev — COMMENTED** [review]
Please remove the console log from the code before merging. The badge collection update functions are missing the step of adding the new earned date, which will lead to a mismatch for the count and earned dates. I have mentioned this to Jae and will create a fix for it. Other than that, the code works as expected.

File: src/controllers/userProfileController.js (hunk 1 above)
Comment (BaileyM96) [reply src/controllers/userProfileController.js]: @ptrpengdev I went through and removed the remaining console logs that you mentioned :) Please let me know if i missed anything else

Comment (ptrpengdev) [reply src/controllers/userProfileController.js]: lgtm. Please do remember to pull the latest code from the development branch and resolve any conflicts. Thanks!

**Comment by Haoxiang310** [comment]
I modified my personalBestMaxHrs and lastWeekTangibleHrs to 40. The code works as intended! 
backend1-PR#619(2)

backend1-PR#619(3)

backend1-PR#619(1)

**Review by 666saofeng — APPROVED** [review]
I have examined this pull request and it functions as expected.
pr619(2)
pr619

**Review by pshereen — CHANGES_REQUESTED** [review]
Backend build is failing.

image

**Pull request merged**